src/js/directives/highlights.js

```diff
@@ -0,0 +1,138 @@
+var rangy = require('rangy');
+var _ = require('lodash');
+var $ = require('jquery');
+
+module.exports = function (app) {
+
+  // returns all leaf text nodes that are descendants of node or are node
+  var getTextNodes = function (node) {
+    if (!node) return [];
+    if (node.nodeType === Node.TEXT_NODE) return [node];
+    
+    // process childs
+    var nodes = [];
+    $(node).contents().each(function (index, el) {
+      nodes = nodes.concat(getTextNodes(el));
+    });
+    return nodes;
+  };
+
+  app.directive('highlightContainer', function () {
+    return {
+      restrict: 'A',
+      transclude: true,
+      replace: true,
+      scope: {},
+      template: '<div ng-transclude></div>',
+      controller: ['$scope', function ($scope) {
+      }]
+    };
+  });
+  
+  app.directive('highlightRoot', function () {
+    return {
+      restrict: 'A',
+      require: '^^highlightContainer',
+      link: function (scope, element, attrs, containerCtrl) {
+        containerCtrl.getRangesRects = function (serializedRanges) {
+          if (!serializedRanges || !serializedRanges.length) return [];
```

Comment:
I don't have a strong opinion about this one. It's clear that length always is an integer and testing for zero with `!` is pretty common, isn't it?